src/psyclone/psyir/nodes/call.py

```diff
@@ -593,3 +600,172 @@ def _location_txt(node):
+                                f"Argument type mismatch of call argument "
+                                f"'{call_arg}' and routine argument "
+                                f"'{routine_arg}'"
+                            )
+
+                    ret_arg_idx_list.append(routine_arg_idx)
+                    named_arg_found = True
+                    break
+
+            if not named_arg_found:
+                # It doesn't match => Raise exception
+                raise CallMatchingArgumentsNotFound(
+                    f"Named argument '{arg_name}' not found"
+                )
+
+            routine_argument_list[routine_arg_idx] = None
+
+        #
+        # Finally, we check if all left-over arguments are optional arguments
+        #
+        for routine_arg in routine_argument_list:
+            routine_arg: DataNode
+
+            if routine_arg is None:
+                continue
+
+            # TODO #759: Optional keyword is not yet supported in psyir.
+            # Hence, we use a simple string match.
+            if ", OPTIONAL" in str(routine_arg.datatype):
+                continue
+
+            raise CallMatchingArgumentsNotFound(
+                f"Argument '{routine_arg.name}' in subroutine"
+                f" '{routine.name}' not handled"
+            )
+
+        return ret_arg_idx_list
+
+    def get_callee(
+                self,
+                ret_arg_match_list: List[int] = None,
+                check_matching_arguments: bool = True):
+        '''
+        Searches for the implementation(s) of the target routine for this Call
+        including argument checks.
+
+        :param ret_arg_match_list: List in which the matching argument
+            indices will be returned
+
+        :returns: the Routine(s) that this call targets.
+        :rtype: list[:py:class:`psyclone.psyir.nodes.Routine`]
+
+        :raises NotImplementedError: if the routine is not local and not found
+            in any containers in scope at the call site.
+        '''
+
+        routine_list = self.get_callees()
+
+        error: Exception = None
+
+        # Search for the routine matching the right arguments
+        for routine in routine_list:
+            routine: Routine
+
+            try:
+                arg_match_list = self.get_argument_routine_match(routine)
+            except CallMatchingArgumentsNotFound as err:
+                error = err
+                continue
+
+            error = None
+
+            if ret_arg_match_list is not None:
+                ret_arg_match_list[:] = arg_match_list
+
+            return routine
+
+        # If we didn't find any routine, return some routine if no matching
+        # arguments have been found.
+        # This is handy for the transition phase until optional argument
+        # matching is supported.
+        if not check_matching_arguments:
+            return routine_list[0]
+
+        if error is not None:
+            raise error
```

Comment:
Perhaps `raise NotImplementedError(blah) from error` here might be nicer?

Reply:
I changed it.